shm_wrapper/shm_wrapper.h

```diff
@@ -16,122 +16,225 @@
-*/
-void device_write_uid (uint64_t dev_uid, process_t process, stream_t stream, uint32_t params_to_write, param_val_t *params);
-
-/*
-Should be called from all processes that want to know current state of the param bitmap (i.e. device handler)
-Blocks on the param bitmap semaphore for obvious reasons
-	- bitmap: pointer to array of 17 32-bit integers to copy the bitmap into
-No return value.
-*/
-void get_param_bitmap (uint32_t *bitmap);
-
-/*
-Should be called from all processes that want to know device identifiers of all currently connected devices
-Blocks on catalog semaphore for obvious reasons
-	- dev_ids: pointer to array of dev_id_t's to copy the information into
-No return value.
-*/
+ * Should be called from every process wanting to write to the device data
+ * Takes care of updating the param bitmap for fast transfer of commands from executor to device handler
+ * Grabs either one or two semaphores depending on calling process and stream requested.
+ * Arguments
+ *    - int dev_ix: device index of the device whose data is being written
+ *    - process_t process: the calling process, one of DEV_HANDLER, EXECUTOR, or NET_HANDLER
+ *    - stream_t stream: the requested block to write to, one of DATA, COMMAND
+ *    - uint32_t params_to_read: bitmap representing which params to be written (nonexistent params should have corresponding bits set to 0)
+ *    - param_val_t params: pointer to array of param_val_t's that is at least as long as highest requested param number
+ *            device data will be written into the corresponding param_val_t's
+ * Returns 0 on success, -1 on failure (specified device is not connected in shm)
+ */
+int device_write (int dev_ix, process_t process, stream_t stream, uint32_t params_to_write, param_val_t *params);
+
+/*
+ * This function is the exact same as the above function, but instead uses the 64-bit device UID to identify
+ * the device that should be written, rather than the device index.
+ */
+int device_write_uid (uint64_t dev_uid, process_t process, stream_t stream, uint32_t params_to_write, param_val_t *params);
+
+/*
+ * Send a sub request to dev_handler for a particular device. Takes care of updating the changed bits.
+ * Should only be called by executor and net_handler
+ * Arguments:
+ *    - uint64_t dev_uid: unique 64-bit identifier of the device
+ *    - process_t process: the calling process (will error if not EXECUTOR or NET_HANDLER)
+ *    - uint32_t params_to_sub: bitmap representing params to subscribe to (nonexistent params should have corresponding bits set to 0)
+ * Returns 0 on success, -1 on failure (unrecognized process, or device is not connect in shm)
+ */
+int place_sub_request (uint64_t dev_uid, process_t process, uint32_t params_to_sub);
+
+/*
+ * Get current subscription requests for all devices. Should only be called by dev_handler
+ * Arguments:
+ *    - uint32_t *sub_map: bitwise OR of the executor and net_handler sub_maps that will be put into this provided buffer
```

Comment:
Clarify that it expects an array of (MAX_DEVICES + 1) uint32_t, where the 0th index is a bitmap indicating which devices require a new sub_request to be sent to

Reply:
It might be clearer (for me at least) if the function declaration was
`void get_sub_requests (uint32_t sub_map[])` to imply an array instead of just a normal pointer

Reply:
👌